pkg/restore/restore.go

```diff
@@ -404,104 +355,66 @@ func (ctx *context) execute() (Result, Result) {
 	// need to set this for additionalItems to be restored
 	ctx.restoreDir = dir
 
+	var (
+		existingNamespaces = sets.NewString()
+		processedResources = sets.NewString()
+	)
+
 	backupResources, err := archive.NewParser(ctx.log, ctx.fileSystem).Parse(ctx.restoreDir)
 	if err != nil {
 		errs.AddVeleroError(errors.Wrap(err, "error parsing backup contents"))
 		return warnings, errs
 	}
 
-	existingNamespaces := sets.NewString()
-
-	for _, resource := range ctx.prioritizedResources {
-		// we don't want to explicitly restore namespace API objs because we'll handle
-		// them as a special case prior to restoring anything into them
-		if resource == kuberesource.Namespaces {
+	// Iterate through an ordered list of resources to restore, checking each one to see if it should be restored.
+	// Note that resources *may* be in this list twice, i.e. once due to being a prioritized resource, and once due
+	// to being in the backup tarball. We can't de-dupe this upfront, because it's possible that items in the prioritized
+	// resources list may not be fully resolved group-resource strings (e.g. may be specfied as "po" instead of "pods"),
+	// and we don't want to fully resolve them via discovery until we reach them in the loop, because it is possible
+	// that the resource/API itself is being restored via a custom resource definition, meaning it's not available via
+	// discovery prior to beginning the restore.
+	//
+	// Since we keep track of the fully-resolved group-resources that we *have* restored, we won't try to restore a
+	// resource twice even if it's in the ordered list twice.
+	for _, resource := range getOrderedResources(ctx.resourcePriorities, backupResources) {
+		// try to resolve the resource via discovery to a complete group/version/resource
+		gvr, _, err := ctx.discoveryHelper.ResourceFor(schema.ParseGroupResource(resource).WithVersion(""))
+		if err != nil {
+			// TODO should we refresh discovery and try again?
```

Comment:
Do we want to resolve this TODO item now, or wait for later? Did it come up often in testing, @skriss?

Reply:
Since I have code to refresh discovery right after restoring custom resource definitions (https://github.com/skriss/velero/blob/refactor-restore-priorities/pkg/restore/restore.go#L456-L464), this didn't come up at all for me, and I think it's probably not necessary.  If we had other resource types beyond CRDs that, upon restore, would necessitate a refresh of discovery, then implementing this TODO might be a better approach than the hardcoded "refresh discovery after restoring CRDs" approach that's implemented now.

I'll remove for now and will need a fresh round of approves :)